apiserver/common/applicationsetstatus.go

```diff
@@ -0,0 +1,120 @@
+	leadershipChecker leadership.Checker
+	st                *state.State
+	getCanModify      GetAuthFunc
+}
+
+// NewApplicationStatusSetter returns a ServiceStatusSetter.
+func NewApplicationStatusSetter(st *state.State, getCanModify GetAuthFunc, leadershipChecker leadership.Checker) *ApplicationStatusSetter {
+	return &ApplicationStatusSetter{
+		leadershipChecker: leadershipChecker,
+		st:                st,
+		getCanModify:      getCanModify,
+	}
+}
+
+// SetStatus sets the status on the service given by the unit in args if the unit is the leader.
+func (s *ApplicationStatusSetter) SetStatus(ctx context.Context, args params.SetStatus) (params.ErrorResults, error) {
+	result := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.Entities)),
+	}
+	if len(args.Entities) == 0 {
+		return result, nil
+	}
+
+	canModify, err := s.getCanModify()
+	if err != nil {
+		return params.ErrorResults{}, err
+	}
+
+	for i, arg := range args.Entities {
+
+		// TODO(fwereade): the auth is basically nonsense, and basically only
+		// works by coincidence. Read carefully.
+
+		// We "know" that arg.Tag is either the calling unit or its service
+		// (because getCanModify is authUnitOrService, and we'll fail out if
+		// it isn't); and, in practice, it's always going to be the calling
+		// unit (because, /sigh, we don't actually use service tags to refer
+		// to services in this method).
+		tag, err := names.ParseTag(arg.Tag)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		if !canModify(tag) {
+			result.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
+		}
+		unitTag, ok := tag.(names.UnitTag)
+		if !ok {
+			// No matter what the canModify says, if this entity is not
+			// a unit, we say "NO".
+			result.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
+		}
+		unitId := unitTag.Id()
+
+		// Now we have the unit, we can get the service that should have been
+		// specified in the first place...
+		serviceId, err := names.UnitApplication(unitId)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		service, err := s.st.Application(serviceId)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+
+		// ...and set the status, conditional on the unit being (and remaining)
+		// service leader.
+		token := s.leadershipChecker.LeadershipCheck(serviceId, unitId)
+
+		// TODO(fwereade) pass token into SetStatus instead of checking here.
+		if err := token.Check(); err != nil {
+			// TODO(fwereade) this should probably be apiservererrors.ErrPerm in certain cases,
+			// but I don't think I implemented an exported ErrNotLeader.
+			// I should have done, though.
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		// TODO(perrito666) 2016-05-02 lp:1558657
+		now := time.Now()
```

Comment:
Should this be clock as well? The bug referenced here is just to use clock https://bugs.launchpad.net/juju/+bug/1558657

Reply:
Yes.

Reply:
It _should_, but this code has a very limited remaining lifespan so there is little point changing it